azure-identity/azure/identity/credentials/providers.py

```diff
@@ -0,0 +1,185 @@
+            raise AuthenticationError("Unexpected response from Active Directory: {}".format(payload))
+
+    @staticmethod
+    def create_config(**kwargs):
+        config = Configuration(**kwargs)
+        config.logging_policy = NetworkTraceLoggingPolicy(**kwargs)
+        # TODO: RetryPolicy should become more straightforwardly configurable
+        config.retry_policy = RetryPolicy(**kwargs)
+        config.retry_policy._retry_on_status_codes = [404, 429] + [
+            x for x in range(500, 600)
+        ]  # pylint:disable=protected-access
+        return config
+
+
+class EnvironmentTokenProvider(ClientSecretTokenProvider):
+    AZURE_CLIENT_ID = "AZURE_CLIENT_ID"
+    AZURE_CLIENT_SECRET = "AZURE_CLIENT_SECRET"
+    AZURE_TENANT_ID = "AZURE_TENANT_ID"
+
+    REQUIRED_VARS = (AZURE_CLIENT_ID, AZURE_CLIENT_SECRET, AZURE_TENANT_ID)
+
+    def __init__(self):
+        # type: () -> None
+        self._undefined_variables = [v for v in self.REQUIRED_VARS if os.environ.get(v) is None]
+        if not self._undefined_variables:
+            super(EnvironmentTokenProvider, self).__init__(
+                client_id=os.environ.get(self.AZURE_CLIENT_ID, ""),
+                secret=os.environ.get(self.AZURE_CLIENT_SECRET, ""),
+                tenant_id=os.environ.get(self.AZURE_TENANT_ID, ""),
+            )
+
+    def get_token(self, scopes):
+        if self._undefined_variables:
+            raise AuthenticationError("environment has no value for {}".format(self._undefined_variables))
+        return super(EnvironmentTokenProvider, self).get_token(scopes)
+
+
+# TODO: support multiple identities?
+class MsiTokenProvider:
+    # https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http
+    _IMDS_ENDPOINT = "http://169.254.169.254/metadata/identity/oauth2/token"
+
+    def __init__(self, config=None, **kwargs):
+        # type: (Optional[Configuration], Mapping[str, Any]) -> None
+        config = config or self.create_config(**kwargs)
+        policies = [config.header_policy, ContentDecodePolicy(), config.logging_policy, config.retry_policy]
+        self._pipeline = Pipeline(transport=RequestsTransport(configuration=config), policies=policies)
+        self._cache = TokenCache()
+
+    def get_token(self, scopes):
+        # type: (Iterable[str]) -> str
+        tokens = self._cache.find(TokenCache.CredentialType.ACCESS_TOKEN, list(scopes))
+        for token in tokens:
+            if all((scope in token["target"] for scope in scopes)):
+                if int(token["expires_on"]) - 300 > int(time()):
+                    return token["secret"]
+
+        # expired token or nothing in the cache => get a token from the endpoint
+        try:
+            request = HttpRequest("GET", self._IMDS_ENDPOINT)
+            # TODO: IMDS allows only one resource per request
+            request.format_parameters({"api-version": "2018-02-01", "resource": " ".join(scopes)})
```

Comment:
>request.format_parameters({"api-version": "2018-02-01", "resource": " ".join(scopes)}) [](start = 12, length = 86)

I believe scopes will have to be modified to call MSI, as MSI currently expects a single resource not a list scopes.  Also I think that  you'll need to strip off the /.Default from the end of scope to get the resource name.